modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+      }
+
+      c_free(buffer);
+    }
+
+    // Phase 3: Release all locks from first node and segment to last node and segment.
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.releaseStatus();
+      }
+    }
+  }
+
+  /*
+    Iterate over each bag in each node. To avoid holding onto locks, we take
+    a snapshot approach, increasing memory consumption but also increasing parallelism.
+    This allows other concurrent, even mutating, operations while iterating,
+    but opens the possibility to iterating over duplicates or missing elements
+    from concurrent operations.
+
+    **Warning:** Breaking from an iterator will leak the snapshot due to issue #6912.
+    It will however leave all Bags in a safe state.
+  */
+  iter these() : eltType {
+    for loc in targetLocales {
+      for segmentIdx in 0..#here.maxTaskPar {
+        // The size of the snapshot is only known once we have the lock.
+        // BUG: c_ptr does not seem to work here and causes a segmentation fault at runtime.
+        var dom : domain(1) = {0..-1};
+        var buffer : [dom] eltType;
+
+        on loc {
+          ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+
+          if segment.acquireIfNonEmpty(STATUS_LOOKUP) {
+            var block = segment.headBlock;
+            while block != nil {
+              for i in 0 .. #block.size {
+                  buffer.push_back(block.elems[i]);
+              }
+              block = block.next;
+            }
+            segment.releaseStatus();
+
+          }
+        }
+
+        // Process this chunk if we have one...
+        for elem in buffer {
+          yield elem;
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind) where tag == iterKind.leader {
+    coforall loc in targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      coforall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+
+        if segment.acquireIfNonEmpty(STATUS_LOOKUP) {
+          // Create a snapshot...
+          var block = segment.headBlock;
+          var bufferSz = segment.nElems.read() : int;
+          var buffer = c_malloc(eltType, bufferSz);
+          var bufferOffset = 0;
+
+          while block != nil {
+            if bufferOffset + block.size > bufferSz {
+              halt("DistributedBag Internal Error: Snapshot attempt with bufferSz(", bufferSz, ") with offset bufferOffset(", bufferOffset + block.size, ")");
+            }
+            __primitive("chpl_comm_array_put", block.elems[0], here.id, buffer[bufferOffset], block.size);
+            bufferOffset += block.size;
+            block = block.next;
+          }
+
+          // Yield this chunk to be process...
+          segment.releaseStatus();
+          yield (bufferSz, buffer);

```

Comment:
This is fine but as a documentation matter you'll have to say that this data structure does not `zip` with rectangular data structures (e.g. forall i,j in zip( myDistributedBag, 1..) )